fs/localfs/shallow_fs.go

```diff
@@ -0,0 +1,118 @@
+package localfs
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/natefinch/atomic"
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/snapshot"
+)
+
+// Helpers to implement storing of "shallow" placeholders for files or
+// directory trees in a restore image. A placeholder for directory d1 is
+// d1.kopia-entry/.kopia-entry. As a result, d1.kopia-entry will stat as
+// a directory and show up nicely in colorized ls output. A placeholder
+// for a file f1 is f1.kopia-entry.
+
+func placeholderPath(path string, et snapshot.EntryType) (string, error) {
+	switch et {
+	case snapshot.EntryTypeFile:
+		return path + SHALLOWENTRYSUFFIX, nil
+	case snapshot.EntryTypeDirectory: // Directories and regular files
+		dirpath := path + SHALLOWENTRYSUFFIX
+		if err := os.MkdirAll(dirpath, os.FileMode(dIRMODE)); err != nil {
+			return "", errors.Wrap(err, "placeholderPath dir creation")
+		}
+
+		return filepath.Join(dirpath, SHALLOWENTRYSUFFIX), nil
+	default:
+		// Shouldn't be used on links or other file types.
+		return "", errors.Errorf("unsupported entry type: %v", et)
+	}
+}
+
+// WriteShallowPlaceholder writes sufficient metadata into the placeholder
+// file associated with path so that it can be roundtripped through
+// snapshot/restore without needing to be realized in the local
+// filesystem.
+// TODO(rjk): Should the placeholder use the complete fs.Entry?
+func WriteShallowPlaceholder(path string, de *snapshot.DirEntry) (string, error) {
+	buffy := &bytes.Buffer{}
+	encoder := json.NewEncoder(buffy)
+
+	if err := encoder.Encode(de); err != nil {
+		return "", errors.Wrapf(err, "json encoding DirEntry")
+	}
+
+	mp, err := placeholderPath(path, de.Type)
+	if err != nil {
+		return "", errors.Wrapf(err, "computing placeholder path: %q", path)
+	}
+
+	// Write the placeholder file.
+	if err := atomic.WriteFile(mp, buffy); err != nil {
+		return "", errors.Wrapf(err, "error writing placeholder to %q", mp)
+	}
+
+	return mp, nil
+}
+
+// ReadShallowPlaceholder returns the decoded ShallowMetadata for path if it exists
+// regardless of the placeholder type.
+func ReadShallowPlaceholder(path string) (*snapshot.DirEntry, error) {
+	originalpresent := false
```

Comment:
i'm worried this will slow down filesystem reads when no placeholders are needed, see my other comment. It would be great to avoid this code path if we can.

Reply:
Ditto. Also addressed in 7f232537.